File: src/main/java/gg/mineral/practice/match/BotMatch.java
package gg.mineral.practice.match;

import org.bukkit.Bukkit;
import org.bukkit.Location;

import gg.mineral.bot.ai.goal.DrinkPotionGoal;
import gg.mineral.bot.ai.goal.EatGappleGoal;
import gg.mineral.bot.ai.goal.MeleeCombatGoal;
import gg.mineral.bot.ai.goal.ReplaceArmorGoal;
import gg.mineral.bot.api.BotAPI;
import gg.mineral.bot.api.configuration.BotConfiguration;
import gg.mineral.bot.api.instance.ClientInstance;
import gg.mineral.practice.PracticePlugin;
import gg.mineral.practice.bots.Difficulty;
import gg.mineral.practice.entity.PlayerStatus;
import gg.mineral.practice.entity.Profile;
import gg.mineral.practice.managers.ArenaManager;
import gg.mineral.practice.managers.MatchManager;
import gg.mineral.practice.managers.ProfileManager;
import gg.mineral.practice.match.data.MatchData;
import gg.mineral.practice.util.PlayerUtil;
import gg.mineral.practice.util.items.ItemStacks;
import lombok.val;

public class BotMatch extends Match {

    private ClientInstance clientInstance;
    private final BotConfiguration config;

    public BotMatch(Profile profile1, BotConfiguration config, MatchData matchData) {
        super(matchData);
        this.profile1 = profile1;
        this.config = config;
        addParicipants(profile1);
    }

    @Override
    public void start() {
        if (noArenas())
            return;

        MatchManager.registerMatch(this);
        val arena = ArenaManager.getArenas().get(getData().getArenaId());
        Location location1 = arena.getLocation1().clone();
        Location location2 = arena.getLocation2().clone();

        setupLocations(location1, location2);

        teleportPlayers(location1, location2);

        this.clientInstance = Difficulty.spawn(config, location2);
        val bukkitPl = Bukkit.getPlayer(config.getUuid());

        if (bukkitPl == null)
            throw new NullPointerException("Fake player is null");

        this.profile2 = ProfileManager
                .getOrCreateProfile(bukkitPl);

        handleOpponentMessages();
        startCountdown();
        addParicipants(profile2);

        prepareForMatch(participants);
    }

    @Override
    public void teleportPlayers(Location location1, Location location2) {
        PlayerUtil.teleport(profile1.getPlayer(), location1);
    }

    @Override
    public void onMatchStart() {
        super.onMatchStart();

        clientInstance.getConfiguration().setPearlCooldown(getData().getPearlCooldown());
        clientInstance.startGoals(new ReplaceArmorGoal(clientInstance), new DrinkPotionGoal(clientInstance),
                new EatGappleGoal(clientInstance),
                new MeleeCombatGoal(clientInstance));
    }

    @Override
    public void end(Profile victim) {
        super.end(victim);

        BotAPI.INSTANCE.despawn(victim.getPlayer().getUniqueId());
    }

    @Override
    public void giveQueueAgainItem(Profile profile) {
        if (BotAPI.INSTANCE.isFakePlayer(profile.getPlayer().getUniqueId()))
            return;

        Bukkit.getServer().getScheduler().runTaskLater(PracticePlugin.INSTANCE,
                () -> profile.getInventory().setItem(profile.getInventory().getHeldItemSlot(),
                        ItemStacks.QUEUE_AGAIN,
                        () -> {
                            if (profile.getPlayerStatus() != PlayerStatus.QUEUEING) {
                                profile.setPlayerStatus(PlayerStatus.QUEUEING);

                                new BotMatch(profile, config,
                                        BotMatch.this.getData()).start();
                            }
                        }),
                20);
    }

    @Override
    public void end(Profile attacker, Profile victim) {
        super.end(attacker, victim);

        BotAPI.INSTANCE.despawn(attacker.getPlayer().getUniqueId());
        BotAPI.INSTANCE.despawn(victim.getPlayer().getUniqueId());
    }

}

File: src/main/kotlin/gg/mineral/practice/match/BotMatch.kt
package gg.mineral.practice.match

import gg.mineral.bot.api.BotAPI
import gg.mineral.bot.api.behaviour.BehaviourTree
import gg.mineral.bot.api.behaviour.node.BTNode
import gg.mineral.bot.api.configuration.BotConfiguration
import gg.mineral.bot.api.instance.ClientInstance
import gg.mineral.bot.impl.behaviour.RootNode
import gg.mineral.practice.PracticePlugin
import gg.mineral.practice.bots.Difficulty
import gg.mineral.practice.entity.PlayerStatus
import gg.mineral.practice.entity.Profile
import gg.mineral.practice.managers.ArenaManager.arenas
import gg.mineral.practice.managers.MatchManager.registerMatch
import gg.mineral.practice.managers.ProfileManager.getOrCreateProfile
import gg.mineral.practice.match.data.MatchData
import gg.mineral.practice.util.PlayerUtil
import gg.mineral.practice.util.items.ItemStacks
import org.bukkit.Bukkit
import org.bukkit.Location
import java.lang.ref.WeakReference

class BotMatch(profile1: Profile, private val config: BotConfiguration, matchData: MatchData) :
    Match(matchData, profile1) {
    private var clientInstance: WeakReference<ClientInstance>? = null

    override fun cleanup() {
        super.cleanup()
        clientInstance = null
    }

    override fun start() {
        if (noArenas()) {
            onError("No arenas available")
            return
        }

        registerMatch(this)
        val arena = arenas[data.arenaId] ?: run {
            onError("Arena not found")
            return
        }
        val location1 = arena.location1.bukkit(world) ?: return
        val location2 = arena.location2.bukkit(world) ?: return

        setupLocations(location1, location2)

        teleportPlayers(location1, location2)

        val instanceFuture = Difficulty.spawn(config, location2)

        instanceFuture.onComplete {
            this.clientInstance = WeakReference(it)
            val player = Bukkit.getPlayer(it.configuration.uuid) ?: error("Player not found")

            getOrCreateProfile(player).let { profile ->
                this.profile2 = profile
                addParticipants(profile)
            }

            handleOpponentMessages()
            startCountdown()

            prepareForMatch(participants)
        }

        instanceFuture.onError {
            participants.forEach { participant ->
                end(participant)
            }
        }
    }

    override fun teleportPlayers(location1: Location, location2: Location) {
        profile1?.let { PlayerUtil.teleport(it, location1) }
    }

    override fun onStart() {
        super.onStart()

        clientInstance?.get()?.let {
            it.configuration.pearlCooldown = data.pearlCooldown
            it.apply {
                behaviourTree = object : BehaviourTree(this) {
                    override val rootNode: BTNode = RootNode(this)
                }
            }
        } ?: onError("Client instance is null")
    }

    override fun end(victim: Profile) {
        super.end(victim)

        BotAPI.INSTANCE.despawn(victim.uuid)
    }

    override fun giveQueueAgainItem(profile: Profile) {
        if (profile.player?.isFake() == true) return

        Bukkit.getServer().scheduler.runTaskLater(
            PracticePlugin.INSTANCE,
            {
                profile.inventory.setItem(
                    profile.inventory.heldItemSlot,
                    ItemStacks.QUEUE_AGAIN,
                    Runnable {
                        if (profile.playerStatus !== PlayerStatus.QUEUEING) {
                            profile.playerStatus = PlayerStatus.QUEUEING

                            BotMatch(
                                profile, config,
                                this@BotMatch.data
                            ).start()
                        }
                    })
            },
            20
        )
    }

    override fun end(attacker: Profile, victim: Profile) {
        super.end(attacker, victim)

        BotAPI.INSTANCE.despawn(attacker.uuid, victim.uuid)
    }
}
